dev/tasks/r/github.nightly.yml

```diff
@@ -0,0 +1,467 @@
+
+      - name: Upload binary artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: r-windows-libarrow
+          path: build/arrow-*.zip 
+
+  r-packages:
+    if: true && !cancelled()
+    needs: [source, windows-cpp]
+    name: {{ '${{ matrix.platform }} ${{ matrix.r_version.r }}' }}
+    runs-on: {{ '${{ matrix.platform }}' }}
+    strategy:
+      fail-fast: false
+      matrix:
+        platform:
+          - windows-latest
+          # This is newer than what CRAN builds on, but Travis is no longer an option for us, so...
+          - macos-10.15
+          # - devops-managed # No M1 until the runner application runs native
+        r_version:
+          - { rtools: 40, r: "4.1" }
+          - { rtools: 42, r: "4.2" }
+    steps:
+      - uses: r-lib/actions/setup-r@v2
+        with:
+          r-version: {{ '${{ matrix.r_version.r }}' }}
+          rtools-version: {{ '${{ matrix.r_version.rtools }}' }}
+          Ncpus: 2
+      {{ macros.github_setup_local_r_repo(is_fork, 'matrix.platform')|indent }}
+      - name: Build Binary
+        shell: Rscript {0}
+        run: |
+          on_windows <- tolower(Sys.info()[["sysname"]]) == "windows"
+
+          # Install dependencies by installing (yesterday's) binary, then removing it
+          install.packages(c("arrow", "cpp11"),
+            type = "binary",
+            repos = c("{{ repo_domain }}", "https://cloud.r-project.org")
+          )
+          remove.packages("arrow")
+
+          dev_repo <- "{{ repo_domain }}"
+
+          # Setup local repo
+          if ({{is_fork|upper}}) {
+            dev_repo <- paste0(
+              ifelse(on_windows, "file:", "file://"),
+              getwd(),
+              "/repo")
+          }
+          options(arrow.dev_repo = dev_repo)
+          
+          # Build
+          Sys.setenv(MAKEFLAGS = paste0("-j", parallel::detectCores()))
+          INSTALL_opts <- "--build"
+          if (!on_windows) {
+            # Windows doesn't support the --strip arg
+            INSTALL_opts <- c(INSTALL_opts, "--strip")
+          }
+
+         
+          install.packages(
+            "arrow",
+            type = "source",
+            repos = dev_repo,
+            INSTALL_opts = INSTALL_opts
+          )
+
+          # Test
+          library(arrow)
+          read_parquet(system.file("v0.7.1.parquet", package = "arrow"))
+      - name: Upload package
+        if: {{ '${{ ' + upload_pkg + ' }}' }}
+        shell: bash
+        env:
+          USER: {{ '${{ secrets.CROSSBOW_NIGHTLIES_USER }}' }}
+          PW: {{ '${{ secrets.CROSSBOW_NIGHTLIES_TOKEN }}' }}
+          VERSION: {{ '${{ needs.source.outputs.version }}' }}
+        run: |
+          REPO_PATH=r$(Rscript -e "cat(contrib.url('', type = 'binary'))")
+          EXT=$(if [[ {{ '${{ matrix.platform }}' }} == windows* ]]; then echo zip; else echo tgz; fi)
+
+          curl -s --fail --show-error -u $USER:$PW -T arrow_$VERSION.$EXT {{ webdav_domain }}/$REPO_PATH/

```

Comment:
How much extra work would it be to move this step to after the `test-*-binary` step? 

It would be nice if we had the whole test run and succeed and only after that upload to the webdav server. This also has the added benefit where the two failure modes of failing to upload (or uploading, but the server not being ready to serve the file when the next step is kicked off) are either super clear or side-stepped entirely.

This would involve adding a bit like Neal original suggested that would imitate the repo on local host during the test stage, but that shouldn't be _so hard_ yeah?

Reply:
Oh actually I hadn't gotten there yet, it looks like you _do_ have the step `update-repo` which does do the updating after the tests run too. So in this case we're actually updating the repo twice? Once her and then once again below?

Or maybe I'm super turned around?

Reply:
Regarding the "only upload once everything finishes" I could do that but I modeled this workflow 1:1 to arrow-r-nightly and it would also mean that we don't upload even if just one of the test jobs (e.g. fedora devel) fails and we don't know if the failure is even related to the build. 

As these are nightlies and are not expected to be 100% stable I think the way it works now is preferable, especially now that we have multiple versions available in the cran repo and if there is an issue you can just use an older one.

Reply:
Also rework incoming...